Pytesseract and fixed ImageMagick cmd

DGrothe-PhD opened the pull request.

mre commented on parser/importer_pytesseract.py:
```diff
@@ -0,0 +1,170 @@
+# Python3
```
Can we use `#!/usr/bin/env python3` here?
Here's an explanation: https://stackoverflow.com/a/19305076/270334

mre commented on parser/importer_pytesseract.py:
```diff
@@ -0,0 +1,170 @@
+
+import os
+
+from PIL import Image
```
Guess this line can be removed. We have the same import below inside the try-catch block. 😉

mre commented on parser/importer_pytesseract.py:
```diff
@@ -0,0 +1,170 @@
+from PIL import Image
+
+BASE_PATH = os.getcwd()
+INPUT_FOLDER = os.path.join(BASE_PATH, "data\\img")
```
These paths will work with Windows.
We could actually make them compatible with Linux, too.
And it's quite easy with pathlib.
https://medium.com/@ageitgey/python-3-quick-tip-the-easy-way-to-deal-with-file-paths-on-windows-mac-and-linux-11a072b58d5f
Would you like to give it a try? 😉

mre commented on parser/importer_pytesseract.py:
```diff
@@ -0,0 +1,170 @@
+    import Image
+import pytesseract
+
+pytesseract.pytesseract.tesseract_cmd = r'C:\Program Files\Tesseract-OCR\tesseract.exe'
```
That's nice, but it only works for other users if they use the same path.

Another way would be to add the tesseract executable in your `PATH` environment variable.
I haven't tried it, but this should work:
```
setx path "%path%;C:\Program Files\Tesseract-OCR\"
```
Then you could remove the `pytesseract.pytesseract.tesseract_cmd` line and it should work for others, too. 😉

mre commented on parser/importer_pytesseract.py:
```diff
@@ -0,0 +1,170 @@
+
+# --- End of pytesseract prep
+
+def prepare_folders():
```
Good idea. Thanks for that.

mre commented on parser/importer_pytesseract.py:
```diff
@@ -0,0 +1,170 @@
+    # 
+    cmd = "convert" + " -rotate " + str(angle)
+    # errors on Windows:
+    # cmd += " '" + input_file + "' '" + output_file + "'"
```
That's interesting! I wonder why the old command fails.
If you have some time, you could print the old and the new command on the commandline and see if there is a difference. Maybe it's because of globbing (a `*` somewhere in the command)?

mre commented on parser/importer_pytesseract.py:
```diff
@@ -0,0 +1,170 @@
+    except RuntimeError as timeout_error:
+        # Tesseract processing is terminated
+        pass
+    #cmd = "tesseract -l deu "
```
These lines can be cleaned up, I guess.

mre commented:
Friendly ping @DGrothe-PhD. 😉

DGrothe-PhD commented:
This was my e-mail to you on 4th of May 2020. I am afraid it got lost somewhere.
[conversation Matthias Github 4th of May 2020.docx](https://github.com/mre/receipt-parser/files/4809435/conversation.Matthias.Github.4th.of.May.2020.docx)

Raw-text copy of that e-mail:
Hi Matthias,

thank you for adding your ideas. I hope I haven't skipped anything.

Regarding #!/usr/bin/env python3 that is something I cannot check: if my script works in Linux or not, or if only the original script will work on some prior versions of Python3.x and/or ImageMagick.
And this is the most important reason why I submitted an extra script instead of amending the original.

Usually, Windows-Python engine does not complain about such a line #!/usr/bin/env python3 at the top. Does Linux always require such a line?

Yes, the extra "from PIL import Image" line can be removed.

However, about pathlib and path-joining stuff, I also cannot check it on Linux. So it would be much better that someone improves my script so that it runs on Linux, and then, I can check if it works on Windows as well.

Perhaps you have an idea why the command failed, I've attached an errorpage.txt which shows the convert.exe errors. To improve readability I added some empty lines and some ### commenting lines. No clue what actually happens with the path string. How for example a ' is inserted inside \IM_MOD_RL_JPG'_.dll or how "C:" is missing there several times although os.getcwd() should pick it up...

The tesseract.exe command path should be generalized in some way. I guess setx is not a good idea though, as there is a 1024-character limit so I can't add further strings from the command line. See also https://stackoverflow.com/questions/17240725/setx-doesnt-append-path-to-system-path-variable on that...

Regards, Daniela

Am 04.05.2020 um 20:56 schrieb Matthias:
>
> @mre approved this pull request.
>
> Thanks for your pull request. 😊
> I've added a few ideas on how to improve the code.
> None of them are necessary to be handled right now, but maybe you have some time and like to have a look. If not, that's also okay of course.
> We could even merge importer_pytesseract.py into importer.py and support all platforms in one script.
> Let me know if you want to continue with this PR or if we should merge your changes already.
>
> In parser/importer_pytesseract.py:
> > @@ -0,0 +1,170 @@
>
> +# Python3
>
> Can we use #!/usr/bin/env python3 here?
> Here's an explanation: https://stackoverflow.com/a/19305076/270334
>
> In parser/importer_pytesseract.py:
> > +# Licensed under the Apache License, Version 2.0 (the "License");
>
> +# you may not use this file except in compliance with the License.
>
> +# You may obtain a copy of the License at
>
> +#
>
> +# http://www.apache.org/licenses/LICENSE-2.0
>
> +#
>
> +# Unless required by applicable law or agreed to in writing, software
>
> +# distributed under the License is distributed on an "AS IS" BASIS,
>
> +# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
>
> +# See the License for the specific language governing permissions and
>
> +# limitations under the License.
>
> +
>
> +
>
> +import os
>
> +
>
> +from PIL import Image
>
> Guess this line can be removed. We have the same import below inside the try-catch block. 😉
>
> In parser/importer_pytesseract.py:
> > +#
>
> +# http://www.apache.org/licenses/LICENSE-2.0
>
> +#
>
> +# Unless required by applicable law or agreed to in writing, software
>
> +# distributed under the License is distributed on an "AS IS" BASIS,
>
> +# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
>
> +# See the License for the specific language governing permissions and
>
> +# limitations under the License.
>
> +
>
> +
>
> +import os
>
> +
>
> +from PIL import Image
>
> +
>
> +BASE_PATH = os.getcwd()
>
> +INPUT_FOLDER = os.path.join(BASE_PATH, "data\\img")
>
> These paths will work with Windows.
> We could actually make them compatible with Linux, too.
> And it's quite easy with pathlib.
> https://medium.com/@ageitgey/python-3-quick-tip-the-easy-way-to-deal-with-file-paths-on-windows-mac-and-linux-11a072b58d5f
> Would you like to give it a try? 😉
>
> In parser/importer_pytesseract.py:
> > +
>
> +from PIL import Image
>
> +
>
> +BASE_PATH = os.getcwd()
>
> +INPUT_FOLDER = os.path.join(BASE_PATH, "data\\img")
>
> +TMP_FOLDER = os.path.join(BASE_PATH, "data\\tmp")
>
> +OUTPUT_FOLDER = os.path.join(BASE_PATH, "data\\txt")
>
> +
>
> +# --- pytesseract prep
>
> +try:
>
> +    from PIL import Image
>
> +except ImportError:
>
> +    import Image
>
> +import pytesseract
>
> +
>
> +pytesseract.pytesseract.tesseract_cmd = r'C:\Program Files\Tesseract-OCR\tesseract.exe'
>
> That's nice, but it only works for other users if they use the same path.
>
> Another way would be to add the tesseract executable in your PATH environment variable.
> I haven't tried it, but this should work:
> setx path "%path%;C:\Program Files\Tesseract-OCR\"
>
> Then you could remove the pytesseract.pytesseract.tesseract_cmd line and it should work for others, too. 😉
>
> In parser/importer_pytesseract.py:
> > +TMP_FOLDER = os.path.join(BASE_PATH, "data\\tmp")
>
> +OUTPUT_FOLDER = os.path.join(BASE_PATH, "data\\txt")
>
> +
>
> +# --- pytesseract prep
>
> +try:
>
> +    from PIL import Image
>
> +except ImportError:
>
> +    import Image
>
> +import pytesseract
>
> +
>
> +pytesseract.pytesseract.tesseract_cmd = r'C:\Program Files\Tesseract-OCR\tesseract.exe'
>
> +the_chosen_lang = 'deu'
>
> +
>
> +# --- End of pytesseract prep
>
> +
>
> +def prepare_folders():
>
> Good idea. Thanks for that.
>
> In parser/importer_pytesseract.py:
> > +
>
> +def rotate_image(input_file, output_file, angle=90):
>
> +    """
>
> +    :param input_file: str
>
> +        Path to image to rotate
>
> +    :param output_file: str
>
> +        Path to output image
>
> +    :param angle: float
>
> +        Angle to rotate
>
> +    :return: void
>
> +        Rotates image and saves result
>
> +    """
>
> +    # 
>
> +    cmd = "convert" + " -rotate " + str(angle)
>
> +    # errors on Windows:
>
> +    # cmd += " '" + input_file + "' '" + output_file + "'"
>
> That's interesting! I wonder why the old command fails.
> If you have some time, you could print the old and the new command on the commandline and see if there is a difference. Maybe it's because of globbing (a * somewhere in the command)?
>
> In parser/importer_pytesseract.py:
> > +
>
> +def run_tesseract(input_file, output_file):
>
> +    """
>
> +    :param input_file: str
>
> +        Path to image to OCR
>
> +    :param output_file: str
>
> +        Path to output file
>
> +    :return: void
>
> +        Runs tesseract on image and saves result
>
> +    """
>
> +    try:
>
> +        tess_to_txt(input_file, output_file)
>
> +    except RuntimeError as timeout_error:
>
> +        # Tesseract processing is terminated
>
> +        pass
>
> +    #cmd = "tesseract -l deu "
>
> These lines can be cleaned up, I guess.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

monolidth commented:
Hi,thanks for your effort. I included this at: 16558037b4f7ea2e10178506a87524fc65b6d306.
